docs/extensions.md

```diff
@@ -0,0 +1,113 @@
+themselves as normal Newspeak classes. Thus, there is no visible difference
+between normal modules and extension modules from the language perspective.
+
+Let's assume we have an extension that provides support for a database.
+It might be used as follows:
+
+```Smalltalk
+public loadExtensionAndConnect = (
+  | (* load a .jar file instead of a .ns file *)
+    moduleClass = system loadModule: 'database.jar'.
+    (* instantiate the module class *)
+    module = moduleClass new. |
+
+  (* connect to database and perform a query *)
+  module connectTo: 'localhost'.
+  ^ module query: 'SELECT * FROM table'
+)
+```
+
+To preserve the semantics of normal Newspeak classes, modules are
+isolated from each other and themselves. Thus, if an extension module is loaded
+multiple times, state that it maintains is keep separated.
+
+## Creating an Extension Module
+
+Extension modules are realized as standard JAR files, which contain a properties
+file that identifies the class implementing the  [`Extension`][1] interface,
+as well as a set of Truffle nodes annotated with [`@Primitive`][2].
+The nodes are used to create a class, which offers them as normal methods.
+
+As a simple example, let's implement a minimal counter primitive:
+
+```Java
+package ext;
+
+@GenerateNodeFactory
+@Primitive(primitive = "inc")
+public class IncPrim extends UnaryExpressionNode {
+  static long cnt = 0;
+
+  @Specialization
+  public Object inc(Object receiver) {
+    cnt += 1;
+    return cnt;
+  }
+}
+```
+
+The `IncPrim` node uses `@Primitive` to define a primitive method named `inc`.
+Its implementation simply increments a counter by one.
+Note that the node is a unary node, which means it takes one argument.
+In Newspeak, the first argument, i.e., the receiver is always present,
+but in this case it is simply ignored by the `inc(.)` method.
+
+Note further the use of `@GenerateNodeFactory`, which is ensures that factory
+class is generated that we will use to create the module class and the methods
+corresponding to these primitives.
+
+The second element we need is an implementation of `Extension`.
+In this example it is simply:
+
+```Java
+package ext;
+
+public class Extension implements som.vm.Extension {
+  @Override
+  public List<NodeFactory<? extends ExpressionNode>> getFactories() {
+    ArrayList<NodeFactory<? extends ExpressionNode>> result = new ArrayList<>();
+    result.add(IncPrimFactory.getFactory());
+    return result;
+  }
+}
+```
+
+An extension needs to provide a list of node factories, which are referring to
+node annotated with `@Primitive`.
+
+The final element we need is the properties file.
+It needs to be named `somns.extension` and be placed in the root of the JAR.
+It defines which class provides the factories, and in this case it is:
+
+```
+class=ext.Extension
+```
+
+With the primitive nodes, the `Extension` class, and the `somns.extension` file,
+we have everything we need in the JAR file.
```

Comment:
Is it possible, without much change, to allow a jar to provide multiple extension classes? Not sure this would ever be helpful but might be worth thinking about.

Reply:
I would not want to go there at this point.
I would stick with the `one file == one module` as we have it for Newspeak modules. Minimal differences between the two is desirable, I think.